Fix gallery synchronization for LastEvil with DLC, and make orgasming avaible.

sis92 opened the pull request:
Modify `GetAnimState` to recognize two naming rule of animation state.
Assume orgasiming is happened in 03_emit where animIndex == 2.

```
[Info   :LoveMachine] UntilReady:Anim_04_End
[Info   :LoveMachine] UntilReady:Anim_03_Emit
[Info   :LoveMachine] UntilReady:Anim_02_During
[Info   :LoveMachine] UntilReady:Anim_01_Start
[Info   :LoveMachine] UntilReady:Animation index: 0


[Info   :LoveMachine] UntilReady:Anim4
[Info   :LoveMachine] UntilReady:Anim3
[Info   :LoveMachine] UntilReady:Anim2
[Info   :LoveMachine] UntilReady:Anim1
[Info   :LoveMachine] UntilReady:Animation index: 0
```

Commit:
[LE] Fix gallery animation for LastEvil with DLC
[LE] Add orgasming state detection

Sauceke commented:
Thank you for the fix. I only have some minor issues, as detailed below.

sis92 commented:
> Thank you for the fix. I only have some minor issues, as detailed below.

I have fixed the issue.

I am interested in developing a plugin for a game developed in Unity 3D, which is available on [Steam](https://store.steampowered.com/app/2897760).

I am unsure how you discovered the bone asset names, scene names, function names, and animation names to detect H-scenes. Would you mind sharing some guidance or the toolchain you used? I would appreciate any help you can provide to get me started.

Sauceke commented:
You can use dnSpy for the function names and Runtime Unity Editor for everything else. They're pretty intuitive to use once you've played around with them for a bit.
Have you checked if it works with `Animation.IsPlaying(name)` instead of the try-catch blocks, by any chance?

sis92 commented:
Thank you for your advice. I am checking `Unity Editor`. That is a fantastic tool!  May I know you did you find the those two functions?

```protected override MethodInfo[] StartHMethods =>
    new[] { AccessTools.Method("EventSceneFramework, Assembly-CSharp:Init"),
    
    };

protected override MethodInfo[] EndHMethods =>
    new[] { AccessTools.Method("EventSceneFramework, Assembly-CSharp:OnClickEnd") };
```

I will try to improve the  try-catch block. 

For testing, do you have any hot-reload tool for it? Or you just restart the game everytime? Any smarter way to do so? Also for the game states, Now I have to play it for a while to trigger the point I want to debug. Any way to directly save & load a snapshot?

Sauceke commented:
I decompiled the game assembly in dnSpy, found the classes that are most likely to manage what I'm looking for, then added a hook for every method in each one whose name even vaguely suggested starting or ending something. You can do the same in RUE, but it doesn't work all the time.
For hot-reloading there is [ScriptEngine](https://github.com/BepInEx/BepInEx.Debug), though I've never used it and don't really like the idea. I think it's best to just use the C# REPL in RUE if you want to debug something. That way you don't have to worry about cleaning up.

Commit:
LastEvil
1. Support Girl vs Girl (with merchant)
2. Compatible with Sex In Battle when you use Temptation card!
3. Fix some animation play, the 32th.
4. Improve efficiency.

sis92 commented:
Thank you! Your tips are very helpful. I have improved a LOT! The feature I most favored is making the battle time sex available! Please check the pull request

sis92 commented:
BTW, there is one thing bothering me. Since it is very difficult to hook the end for the battle scene. I want to end H when I detect animation is null.

I can detect it now. But EndH is not visible from inherited GameAdapter. So I tried a function call  StopAllCoroutines(); 

https://github.com/sis92/LoveMachine/blob/1af4f64d5766d53135635758805049f053fbe38b/src/LoveMachine.LE/LastEvilGame.cs#L88C16-L88C37

However, it doesn't work as expected.

`[Error  :LoveMachine] GetAnimState: Animation is null, endH
[Error  :LoveMachine] GetAnimState: Animation is null, endH
[Error  :LoveMachine] GetAnimState: Animation is null, endH
[Error  :LoveMachine] GetAnimState: Animation is null, endH
[Error  :LoveMachine] GetAnimState: Animation is null, endH`

Commit:
LastEvil

Add myEndH method and update animation handling

sis92 commented:
Any tips for IL2CPP games? UnityExplorer,  RUE failed in that game. Il2CPP is very problematic.

I noticed that you have developed client for Ricca, which also use IL2CPP, BepInEx 6.0. How did you find scene objects ?

Sauceke commented:
I just realized I didn't hit finish review and you couldn't see all my review comments. (Thanks github, bloody helpful as always.) I'll just start a new one.

Sauceke commented on src/LoveMachine.LE/LastEvilGame.cs:
```diff
@@ -71,10 +84,42 @@ protected override Animator GetFemaleAnimator(int girlIndex) =>
+                length = state.length;
+                speed = state.speed;
+
+                //Logger.LogInfo("GetAnimState: Animation state found: " + state.name +
```
Please remove this if not needed or uncomment if needed.

Sauceke commented on src/LoveMachine.LE/LastEvilGame.cs:
```diff
@@ -71,10 +84,42 @@ protected override Animator GetFemaleAnimator(int girlIndex) =>
+                //    " length: " + length +
+                //    " speed: " + speed);
+            }
+            else
```
You can use early return instead.

sis92 replied:
I still use if-else. VS tell me the those three parameters need to be assigned in all possible branches. I cannot just simply return and leave the value unassigned.

Sauceke replied:
``` C#
            if (state != null)
            {   
                normalizedTime = state.time / state.length;
                length = state.length;
                speed = state.speed;
                return;
            }
            // No valid animation found, set default values
            normalizedTime = 0f;
            length = 1f;
            speed = 1f;
```

sis92 replied:
May I know why this is better than if-else? Not very familiar with Csharp

Sauceke replied:
It may not look much different as-is but imagine that you have another if statement inside the else block. Then another inside the else block of that. It can get out of hand pretty fast. A static analyzer will actually catch this and always recommend converting to early return when possible.

Sauceke commented on src/LoveMachine.LE/LastEvilGame.cs:
```diff
@@ -71,10 +84,42 @@ protected override Animator GetFemaleAnimator(int girlIndex) =>
+            {
+                foreach (var anim in animation)
+                {
+                    Logger.LogInfo("GetAnimState Not Hit:" + ((AnimationState)anim).name);
```
Do we need this?

Sauceke commented on src/LoveMachine.LE/LastEvilGame.cs:
```diff
@@ -89,13 +134,45 @@ protected override GameObject GetFemaleRoot(int girlIndex) =>
+
+            // Try to cast eventSceneFramework to the correct type, like Unity.GameObject
+            UnityEngine.Object obj = eventSceneFramework as UnityEngine.Object;
+            root = obj.name;
```
Uh... it's only a coincidence this works. The plugin can't find the bones using this path, so it searches the entire scene, which is obviously not ideal. This is partly my fault, I should have added some logging to make this clear.
The old path for root (`EventSceneFramework/Root/Entities`) should be alright, since the `AnimCombineEventer` is also under that. But there is an alternative name for the female character which I hadn't noticed before. I think `GetFemaleRoot` is supposed to be `GameObject.Find(root + "/Succubus") ?? FindDeepChildrenByPath(GameObject.Find(root), "Female").FirstOrDefault()` or something similar.

sis92 replied:
This is critical for the feature `In battle H scene detection`. When the temptation card is used, an `AnimCombineEventer` is created to play the animation, but there is no `EventSceneFramework`.

sis92 replied:
I changed the GetFemaleRoot as suggested. Yeah, I don't understand why it worked previously.

Sauceke replied:
The path is still broken if the `EventSceneFramework` exists. Branch based on `eventSceneFramework.GetType().Name`? The new path if it's AnimCombineEventer, the old one if it isn't?

sis92 replied:
No, the path root should be the instance name but not type name, isn't it? I suppose `eventSceneFramework.GetType().Name` is the name of Class.

For regular H scene, it looks like
```
- EventSceneFramework (EventSceneFramework Class)
- - Entity 
- - - Sccubus
.....
```

For Temptation, it looks like
```
- AnimCombineEventer_Hmn_ (AnimCombineEventer Class)
- - Female
- - Male
```

Sauceke replied:
Yup, that's what I'm saying. You can put both root paths in a ternary and select the second if `eventSceneFramework.GetType().Name == "AnimCombineEventer"`, otherwise the first.
And then you wouldn't even need deep child search for `GetFemaleRoot`, it could just be `GameObject.Find(root + "/Succubus") ?? GameObject(root + "/Female")`. Or you could even make `root` the Transform itself instead of a string, and have `root.Find("Succubus") ?? root.Find("Female")`, if you like that better.

Sauceke commented on src/LoveMachine.LE/LastEvilGame.cs:
```diff
@@ -71,10 +84,42 @@ protected override Animator GetFemaleAnimator(int girlIndex) =>
+                }
+                // No valid animation found, set default values
+                normalizedTime = 0f;
+                length = 0f;
```
I'm not sure how well the code responds to having 0 length, IMO best to make this 1.

sis92 replied:
fixed

Sauceke commented on src/LoveMachine.LE/LastEvilGame.cs:
```diff
@@ -37,16 +41,25 @@ public class LastEvilGame : GameAdapter
         protected override MethodInfo[] EndHMethods =>
-            new[] { AccessTools.Method("EventSceneFramework, Assembly-CSharp:OnClickEnd") };
+            new[] { AccessTools.Method("EventSceneFramework, Assembly-CSharp:OnClickEnd"),
+            AccessTools.Method(typeof(LastEvilGame), nameof(myEndH))};
```
Have you tried `AnimCombineEventer.SetEnd`?

sis92 replied:
Yeah, i tried that. But it stops animation sequence early. According to dnspy. The SetEnd will start a corountine to play animation 03 and 04. So that is not the actual end.

Sauceke commented:
> Any tips for IL2CPP games? UnityExplorer, RUE failed in that game. Il2CPP is very problematic.

https://github.com/yukieiji/UnityExplorer/tree/v4.12.1
Il2cpp is the worst, it makes most of the code completely opaque. Only thing you can do is poke it with a stick repeatedly and see what comes out.

Commit:
LastEvil Plugin
Remove unused logger and comments
Improve robustness of GetFemaleRoot.

sis92 reviewed (commented):
All comments have been addressed.

Sauceke commented:
https://github.com/Sauceke/LoveMachine/pull/187#discussion_r2094568527
https://github.com/Sauceke/LoveMachine/pull/187#discussion_r2094572770

Sauceke commented:
Happy to merge once the remaining 2 threads are resolved.

Sauceke commented:
Since this is almost done, I'll merge it and do the remaining changes myself.
Thanks for all the improvements.

sis92 commented:
Thank you! Sorry for the late reply. I haven't touch this project quite
yet. Good to know you may integrate it!

On Sat, Aug 9, 2025, 7:06 AM Sauceke ***@***.***> wrote:

> Merged #187 <https://github.com/Sauceke/LoveMachine/pull/187> into mastur.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/Sauceke/LoveMachine/pull/187#event-19067193827>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/BPPKSQQSK6ZTPLLGGCCJQGD3MXI5DAVCNFSM6AAAAAB5KAAF62VHI2DSMVQWIX3LMV45UABCJFZXG5LFIV3GK3TUJZXXI2LGNFRWC5DJN5XDWMJZGA3DOMJZGM4DENY>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>
